src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -179,12 +135,128 @@ int MPL_gpu_free(void *ptr)
     return MPL_ERR_GPU_INTERNAL;
 }
 
-int MPL_gpu_init()
+int MPL_gpu_init(int local_size, int local_rank, int need_thread_safety)
 {
-    return MPL_SUCCESS;
+    int i, mpl_err = MPL_SUCCESS;
+
+    node_local_size = local_size;
+    self_local_rank = local_rank;
+    cuda_ipc_handle_trees = MPL_malloc(sizeof(MPL_gavl_tree_t) * local_size, MPL_MEM_OTHER);
+    assert(cuda_ipc_handle_trees != NULL);
+    for (i = 0; i < local_size; ++i) {
+        mpl_err =
+            MPL_gavl_tree_create(sizeof(gpu_ipc_handle_obj_s), need_thread_safety,
+                                 &cuda_ipc_handle_trees[i]);
+        if (mpl_err != MPL_SUCCESS) {
+            MPL_gpu_finalize();
+            break;
+        }
+    }
+    return mpl_err;
 }
 
 int MPL_gpu_finalize()
 {
+    if (cuda_ipc_handle_trees) {
+        int i;
+        for (i = 0; i < node_local_size; ++i)
+            if (cuda_ipc_handle_trees[i])
+                MPL_gavl_tree_free(cuda_ipc_handle_trees[i], gpu_ipc_handle_free);
+    }
+    MPL_free(cuda_ipc_handle_trees);
+    return MPL_SUCCESS;
+}
+
+static void gpu_ipc_handle_free(void *handle_obj)
+{
+    gpu_ipc_handle_obj_s *handle_obj_ptr = (gpu_ipc_handle_obj_s *) handle_obj;
+    cudaIpcCloseMemHandle((void *) ((char *) handle_obj_ptr - handle_obj_ptr->offset));
```

Comment:
This seems convoluted.  What are you trying to do here?  Are you trying to get the first field of `handle_obj`, i.e., `remote_base_addr`?  Please use more explicit code, so it's easier for future code maintainers to understand it.

Reply:
`handle_obj` might be inserted into AVL, so it should be alive until tree is freed. It has to be malloc instead of local stack variable.